Fix `extrema(x; dims)` for inputs with `NaN/missing`

N5N3 opened the pull request:
The original goal of this PR is to Fix #43599. But I find some more cases whose result is inconsistent with the definition.
The behaviour changes are summarized as follows:
- [x] `extrema(x; dims)` now respects `-0,0/0.0` (i.e. `extrema([-0.0;0.0]; dims = 1) == [(-0.0, 0.0)]`)
- [x] `extrema(x; dims)` now respects `NaN`
- [x] `extrema(x; dims)` can handle inputs with `missing`
- [x] `extrema(x; dims)` disallows empty reduction like `minimun` and `maximum`.
- [x] `minimum(x)` and `maximum(x)` give correct result if the input contains both `NaN` and `missing`

The last one is used in test so I fix it in this PR by limiting the related optimizaiton to `Float` only cases. (see 76637c7)

tkf committed:
Define `extrema` using `mapreduce`; support `init`

tkf committed:
Add NEWS

tkf committed:
Fix a typo

tkf committed:
Reword a comment slightly

tkf committed:
Merge branch 'master' into extrema

tkf committed:
Apply suggestions from code review

Co-authored-by: Milan Bouchet-Valat <nalimilan@club.fr>

tkf committed:
Fixup docstrings

tkf committed:
Add simple tests for Core.Compiler.extrema

tkf committed:
Name the output tuple elements (mn, mx)

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into extrema

tkf committed:
Merge branch 'master' into extrema

tkf committed:
Update base/multidimensional.jl

tkf committed:
Fix compat annotations

tkf committed:
Improve extrema docstring

Co-authored-by: Tim Holy <tim.holy@gmail.com>

tkf committed:
Fix tests for SparseArrays

N5N3 commented:
Downloads error on linux32 should be unrelated.

N5N3 commented:
Performance seems better if the unroll size is increased to 8 (from 4).
I drop the original `_fast` optimization as it might be costy if `f !== identity`.

The newest Benchmark:
```julia
using Random, BenchmarkTools                   
Random.seed!(0);
a = rand(128,128);                  # 1.7.1        This PR
@btime extrema($a, dims = 1);       #21.700 μs  --> 11.800 μs
@btime extrema($a, dims = 2);       #20.800 μs  --> 10.300 μs
@btime extrema($a, dims = (1,2));   #19.600 μs  -->  8.267 μs
@btime extrema($a);                 #37.200 μs  -->  7.850 μs
a = rand(-20:20,128,128);                                            
@btime extrema($a, dims = 1);       #20.400 μs  --> 4.571 μs
@btime extrema($a, dims = 2);       #19.800 μs  --> 9.900 μs
@btime extrema($a, dims = (1,2));   #17.600 μs  --> 3.587 μs
@btime extrema($a);                 #2.5220 μs  --> 3.150 μs
a = rand(Float32,128,128);                                            
@btime extrema($a, dims = 1);       #20.600 μs  --> 11.100 μs
@btime extrema($a, dims = 2);       #20.300 μs  --> 13.200 μs
@btime extrema($a, dims = (1,2));   #18.900 μs  --> 8.500 μs
@btime extrema($a);                 #37.200 μs  --> 8.067 μs
           
Random.seed!(0);
a = rand(128,128);                  # 1.7.1     This PR
@btime minimum($a, dims = 1);       #32.200 μs  --> 8.700 μs
@btime minimum($a, dims = 2);       #6.040 μs  --> 6.060 μs
@btime minimum($a, dims = (1,2));   #8.633 μs  --> 6.240 μs
@btime minimum($a);                 #7.800 μs  --> 5.417 μs
a = rand(-20:20,128,128);                                            
@btime minimum($a, dims = 1);       #3.763 μs  --> 3.650 μs
@btime minimum($a, dims = 2);       #2.922 μs  --> 3.062 μs
@btime minimum($a, dims = (1,2));   #5.550 μs  --> 2.767 μs
@btime minimum($a);                 #4.714 μs  --> 1.740 μs
a = rand(Float32,128,128);                                            
@btime minimum($a, dims = 1);       #32.000 μs --> 10.200 μs
@btime minimum($a, dims = 2);       #3.487 μs  --> 3.475 μs
@btime minimum($a, dims = (1,2));   #8.333 μs  --> 8.000 μs
@btime minimum($a);                 #7.525 μs  --> 7.175 μs
```

N5N3 commented:
The unroll size is further tuned for better performace.
The latest Benchmark:
```julia           
using Random, BenchmarkTools                   
Random.seed!(0);
a = rand(128,128);                  # 1.7.1        This PR
@btime extrema($a, dims = 1);       #21.700 μs  --> 10.500 μs
@btime extrema($a, dims = 2);       #20.800 μs  --> 10.300 μs
@btime extrema($a, dims = (1,2));   #19.600 μs  -->  7.575 μs
@btime extrema($a);                 #37.200 μs  -->  7.150 μs
a = rand(-20:20,128,128);                                            
@btime extrema($a, dims = 1);       #20.400 μs  --> 4.571 μs
@btime extrema($a, dims = 2);       #19.800 μs  --> 9.900 μs
@btime extrema($a, dims = (1,2));   #17.600 μs  --> 3.587 μs
@btime extrema($a);                 #2.5220 μs  --> 3.150 μs
a = rand(Float32,128,128);                                            
@btime extrema($a, dims = 1);       #20.600 μs  --> 8.500 μs
@btime extrema($a, dims = 2);       #20.300 μs  --> 13.200 μs
@btime extrema($a, dims = (1,2));   #18.900 μs  --> 5.550 μs
@btime extrema($a);                 #37.200 μs  --> 5.150 μs
           
Random.seed!(0);
a = rand(128,128);                  # 1.7.1        This PR
@btime minimum($a, dims = 1);       #32.200 μs --> 8.167 μs
@btime minimum($a, dims = 2);       #6.040 μs  --> 5.483 μs
@btime minimum($a, dims = (1,2));   #8.633 μs  --> 5.767 μs
@btime minimum($a);                 #7.800 μs  --> 5.350 μs
a = rand(-20:20,128,128);                                            
@btime minimum($a, dims = 1);       #3.763 μs  --> 3.250 μs
@btime minimum($a, dims = 2);       #2.922 μs  --> 2.578 μs
@btime minimum($a, dims = (1,2));   #5.550 μs  --> 2.244 μs
@btime minimum($a);                 #4.714 μs  --> 1.800 μs
a = rand(Float32,128,128);                                            
@btime minimum($a, dims = 1);       #32.000 μs --> 7.275 μs
@btime minimum($a, dims = 2);       #3.487 μs  --> 2.944 μs
@btime minimum($a, dims = (1,2));   #8.333 μs  --> 3.825 μs
@btime minimum($a);                 #7.525 μs  --> 3.450 μs
```

N5N3 commented:
The unroll size is further tuned for better performace.
The latest Benchmark:
```julia           
using Random, BenchmarkTools                   
Random.seed!(0);
a = rand(128,128);                  # 1.7.1        This PR
@btime extrema($a, dims = 1);       #21.700 μs  --> 10.500 μs
@btime extrema($a, dims = 2);       #20.800 μs  --> 10.300 μs
@btime extrema($a, dims = (1,2));   #19.600 μs  -->  7.575 μs
@btime extrema($a);                 #37.200 μs  -->  7.150 μs
a = rand(-20:20,128,128);                                            
@btime extrema($a, dims = 1);       #20.400 μs  --> 4.571 μs
@btime extrema($a, dims = 2);       #19.800 μs  --> 9.900 μs
@btime extrema($a, dims = (1,2));   #17.600 μs  --> 3.587 μs
@btime extrema($a);                 #2.5220 μs  --> 3.150 μs
a = rand(Float32,128,128);                                            
@btime extrema($a, dims = 1);       #20.600 μs  --> 8.500 μs
@btime extrema($a, dims = 2);       #20.300 μs  --> 13.200 μs
@btime extrema($a, dims = (1,2));   #18.900 μs  --> 5.550 μs
@btime extrema($a);                 #37.200 μs  --> 5.150 μs
           
Random.seed!(0);
a = rand(128,128);                  # 1.7.1        This PR
@btime minimum($a, dims = 1);       #32.200 μs --> 8.167 μs
@btime minimum($a, dims = 2);       #6.040 μs  --> 5.483 μs
@btime minimum($a, dims = (1,2));   #8.633 μs  --> 5.767 μs
@btime minimum($a);                 #7.800 μs  --> 5.350 μs
a = rand(-20:20,128,128);                                            
@btime minimum($a, dims = 1);       #3.763 μs  --> 3.250 μs
@btime minimum($a, dims = 2);       #2.922 μs  --> 2.578 μs
@btime minimum($a, dims = (1,2));   #5.550 μs  --> 2.244 μs
@btime minimum($a);                 #4.714 μs  --> 1.800 μs
a = rand(Float32,128,128);                                            
@btime minimum($a, dims = 1);       #32.000 μs --> 7.275 μs
@btime minimum($a, dims = 2);       #3.487 μs  --> 2.944 μs
@btime minimum($a, dims = (1,2));   #8.333 μs  --> 3.825 μs
@btime minimum($a);                 #7.525 μs  --> 3.450 μs
```

N5N3 commented:
The unroll size is further tuned for better performace.
The latest Benchmark:
```julia           
using Random, BenchmarkTools                   
Random.seed!(0);
a = rand(128,128);                  # 1.7.1        This PR
@btime extrema($a, dims = 1);       #21.700 μs  --> 10.900 μs
@btime extrema($a, dims = 2);       #20.800 μs  --> 10.300 μs
@btime extrema($a, dims = (1,2));   #19.600 μs  -->  6.340 μs
@btime extrema($a);                 #37.200 μs  -->  5.967 μs
a = rand(-20:20,128,128);                                            
@btime extrema($a, dims = 1);       #20.400 μs  --> 4.571 μs
@btime extrema($a, dims = 2);       #19.800 μs  --> 9.900 μs
@btime extrema($a, dims = (1,2));   #17.600 μs  --> 3.587 μs
@btime extrema($a);                 #2.5220 μs  --> 3.150 μs
a = rand(Float32,128,128);                                            
@btime extrema($a, dims = 1);       #20.600 μs  --> 8.733 μs
@btime extrema($a, dims = 2);       #20.300 μs  --> 12.900 μs
@btime extrema($a, dims = (1,2));   #18.900 μs  --> 5.100 μs
@btime extrema($a);                 #37.200 μs  --> 4.714 μs
           
Random.seed!(0);
a = rand(128,128);                  # 1.7.1        This PR
@btime minimum($a, dims = 1);       #32.200 μs --> 7.675 μs
@btime minimum($a, dims = 2);       #6.040 μs  --> 5.533 μs
@btime minimum($a, dims = (1,2));   #8.633 μs  --> 4.629 μs
@btime minimum($a);                 #7.800 μs  --> 4.186 μs
a = rand(-20:20,128,128);                                            
@btime minimum($a, dims = 1);       #3.763 μs  --> 3.250 μs
@btime minimum($a, dims = 2);       #2.922 μs  --> 2.500 μs
@btime minimum($a, dims = (1,2));   #5.550 μs  --> 2.156 μs
@btime minimum($a);                 #4.714 μs  --> 1.760 μs
a = rand(Float32,128,128);                                            
@btime minimum($a, dims = 1);       #32.000 μs --> 6.420 μs
@btime minimum($a, dims = 2);       #3.487 μs  --> 2.878 μs
@btime minimum($a, dims = (1,2));   #8.333 μs  --> 3.575 μs
@btime minimum($a);                 #7.525 μs  --> 3.175 μs
```

N5N3 commented:
Just notice that we already have #36265, which should be able to fix the original concern of this PR. (And apparently, `init` support is useful sometime.)
So the main improvement of this PR is a better vectorlizable `mapreduce_impl` for `min/max/extrema`:
1. Perform `nan` check via `Base._anynan`, before `min/max(x, y)`
    This ensures better vectorization:
    ```julia
    julia> @code_llvm debuginfo = :none Base._anynan(ntuple(Float64,8))
    ; Function Attrs: uwtable
    define i8 @julia__anynan_16819([8 x double]* nocapture nonnull readonly align 8 dereferenceable(64) %0) #0 {
    top:
      %1 = getelementptr inbounds [8 x double], [8 x double]* %0, i64 0, i64 4
      %2 = bitcast [8 x double]* %0 to <4 x double>*
      %3 = load <4 x double>, <4 x double>* %2, align 8
      %4 = bitcast double* %1 to <4 x double>*
      %5 = load <4 x double>, <4 x double>* %4, align 8
      %6 = fcmp uno <4 x double> %3, %5
      %7 = call i1 @llvm.vector.reduce.or.v4i1(<4 x i1> %6)
      %8 = zext i1 %7 to i8
      ret i8 %8
    }
    ```
   And thus we can do `min/max` safely via `ifelse(signbit(x-y), ...)`
2. The unroll-size of loop is better tuned for different input length with a reused `simd_kernal`.

There was no `mapreduce_impl` for `extrema`, so I only bench `min` (`max` should be similar) here:
```julia
using Random, BenchmarkTools
Random.seed!(0);
a = randn(4096);
for n = 1:12
    print(1<<n,':')
    @btime Base.mapreduce_impl(i->i,min, $a, 1, (1<<$n))     
end
```
on 1.7.1
```julia
2:  8.600 ns (0 allocations: 0 bytes)
4:  11.311 ns (0 allocations: 0 bytes)
8:  17.417 ns (0 allocations: 0 bytes)
16:  31.859 ns (0 allocations: 0 bytes)
32:  61.803 ns (0 allocations: 0 bytes)
64:  121.657 ns (0 allocations: 0 bytes)
128:  241.162 ns (0 allocations: 0 bytes)
256:  481.026 ns (0 allocations: 0 bytes)
512:  601.685 ns (0 allocations: 0 bytes)
1024:  837.333 ns (0 allocations: 0 bytes)
2048:  1.300 μs (0 allocations: 0 bytes)
4096:  2.233 μs (0 allocations: 0 bytes)
```
after c148ec9
```julia
2:  7.500 ns (0 allocations: 0 bytes)
4:  10.210 ns (0 allocations: 0 bytes)
8:  15.631 ns (0 allocations: 0 bytes)
16:  17.117 ns (0 allocations: 0 bytes)
32:  21.765 ns (0 allocations: 0 bytes)
64:  32.261 ns (0 allocations: 0 bytes)
128:  47.118 ns (0 allocations: 0 bytes)
256:  75.128 ns (0 allocations: 0 bytes)
512:  111.613 ns (0 allocations: 0 bytes)
1024:  187.391 ns (0 allocations: 0 bytes)
2048:  342.793 ns (0 allocations: 0 bytes)
4096:  665.823 ns (0 allocations: 0 bytes)
```
This PR give a 2~4x speed up for `length > 16`

N5N3 commented:
@tkf is there a strong objection for #36265?
I think it should be good to merge that first, and rebase this one.

JeffBezanson commented:
I like #36265, but it doesn't seem to solve the issue with `missing`?

N5N3 commented:
I missed the following:
https://github.com/JuliaLang/julia/blob/5669d6349569ff83394208609c1d5c4b2f86fa17/base/multidimensional.jl#L1754-L1762

So it seems that PR also didn't solve `NaN`.

But the `_DupY(f)` design is good, and I want to use it in this PR.

tkf commented:
> @tkf is there a strong objection for #36265?

I mentioned it in https://github.com/JuliaLang/julia/pull/36265#issuecomment-967784155 but it looks like I broke some improvements in SparseArrays (IIRC introduced in #37429).

N5N3 commented:
Sparse input with `NaN` is also fixed. (By reusing most of the general Sparse reducedim with a little optimizaiton.)
Speedtest shows that it's faster than dense case for real sparse input. So I guess further optimization is not needed at current stage.

For API clean, I borrowed some code from #36265.
Now most `extrema` is relied on `mapreduce` like `minimum`.
And thus `init` is also supported. (`_extrema_dims` and `_extrema_itr` is totally removed)

N5N3 commented:
To conclude the newest commits:
1. ce42aec8f4ef3d7369028f91db9f67ce3349cad0 is a further attemp of #36265 to achieve uniform implementaion. (`_extrema_dims` and `_extrema_itr` is totally removed). Now most `extrema` is `mapreduce` based and thus `init` is supported. (related test is also added)
2. 6e8abdb72e8ed4940c3bfc156db1445c2c5bccd2 simplfies the initialization of `minimum`/`maximum` with `dims`. And add that for `extrema`.
3. 4626bab35b3bdb315ad4d1799105643371625d23 and 5429bb6abcde0969e1e262700467d0d83cb32efb are for performance.
4. 79eefa62a6232532a51a78d3404fce8628d0a00d fixes Sparse input with `NaN`  by reusing most of the general Sparse reduce(dim). Speedtest shows that it's faster than dense case for real sparse input. So I guess further optimization is not needed at the current stage.
5. f83ed41cb1720ce856e0f2d43c2493e8fffadf4a adds more test for `NaN/missing`

BTW, should we document `extrema!`? It was implemented but not exported and documented (and possibly not tested). But it seems reasonable to make it public.

N5N3 commented:
To conclude the newest commits:
1. 74919dca0ca6f842e0bf6387b60633aabe9c313a is a further attemp of #36265 to achieve uniform implementaion. (`_extrema_dims` and `_extrema_itr` is totally removed). Now most `extrema` is `mapreduce` based and thus `init` is supported. (related test is also added)
2. c12e7122e77b95b71c4fa7f9dd61adf2549a6d05 simplfies the initialization of `minimum`/`maximum` with `dims`. And add that for `extrema`.
3. 54baf1cfac2dd8ed71c7f74affe99dce6d92eaed and ff6b1fa2c94e5212d0186818ed000ea64ba263ed are for performance.
4. a7c3ff7ab52fdf6699e61d9656778ed11df02075 fixes Sparse input with `NaN`  by reusing most of the general Sparse reduce(dim). Some speed test:
    ```julia
    using SparseArrays, BenchmarkTools, Random
    Random.seed!(0);
    a = sprandn(128,128,0.3);
    @btime extrema($a, dims = 1); #1.7.1:7.450μs    PR:7.650 μs
    @btime extrema($a, dims = 2); #1.7.1:264.900μs  PR:11.300 μs
    @btime extrema($a, dims = :); #1.7.1:24.200μs   PR:1.670 μs
    v = sprandn(4096,0.3);
    @btime extrema($v);           #1.7.1:6.160 μs   PR:461.929 ns
    ```
5. ece2a096617e2b016e194c28bcc46891959eb20b adds more test for `NaN/missing`

BTW, should we document `extrema!`? It was implemented but not exported and documented (and possibly not tested). But it seems reasonable to make it public.

tkf commented:
Thanks a lot for picking #36265 up. Do you mind actually merging the branch of #36265 (https://github.com/tkf/julia/tree/extrema)? (Or rebase on top of it, if you like.) It'd add contributors there and also makes edits clearer. For example, 74919dca0ca6f842e0bf6387b60633aabe9c313a seems to remove some while lines in the docstring which would be nice to keep: https://github.com/JuliaLang/julia/pull/43604/commits/74919dca0ca6f842e0bf6387b60633aabe9c313a#diff-97ad8b900c30b39398c32d78754c404f4a6df9a153d45284b2c11ab54837deb9R789-R799

> BTW, should we document `extrema!`? It was implemented but not exported and documented (and possibly not tested). But it seems reasonable to make it public.

Yes, now that `extrema!` is derived in a same way as `minimum!` etc., it seems that it respects the proper ordering (compared to the previous implementation which had some broken edge cases). I agree it makes sense to make it public now.

tkf commented on base/mpfr.jl:
```diff
@@ -706,6 +706,15 @@ function min(x::BigFloat, y::BigFloat)
     return z
 end
 
+function Base._extrema_rf(x::NTuple{2,BigFloat}, y::NTuple{2,BigFloat})
+    (x1, x2), (y1, y2) = x, y
+    isnan(x1) && return x
+    isnan(y1) && return y
+    z1 = x1 < y1 || signbit(x1) > signbit(y1) ? x1 : y1
+    z2 = x2 < y2 || signbit(x2) > signbit(y2) ? y2 : x2
+    z1, z2
+end
```
Is it for "undoing" the branch-free optimization for `AbstractFloat`?

https://github.com/JuliaLang/julia/blob/ece2a096617e2b016e194c28bcc46891959eb20b/base/reduce.jl#L785-L793

If so, how about just calling the generic method `invoke(_extrema_rf, Tuple{Any,Any}, x, y)`? If the computation of `z1` and `z2` is correct, why not do this in `min` and `max` as well? Also, it's better to put some comments on why certain optimization is done and why it is correct.

I'm not familiar with `BigFloat` so I don't know if it's OK. It may be better to open a separate PR that focuses on optimizing `min`/`max`.

N5N3 replied:
See #41709, `BigFloat`'s `min/max` do allocation for non-nan inputs.
But `extrema` for `BigFloat` don't create new `BigFloat` before this PR.
More allocations might be considered as a regression, so I add it in `mpfr.jl` (for  bootstrap)

N5N3 replied:
> Do you mind actually merging the branch of #36265 (https://github.com/tkf/julia/tree/extrema)?
 
I dont mind, the easizer way for me might be cherry pick the commits from #36265.
But I got lost when I look into git histry... (So I choosed to copy the code directly...)

N5N3 replied:
Comments added.

tkf replied:
The right way to fix this seems to make `min`/`max` non-allocating. It's better to clarify the desired semantics of `min`/`max` first. If it really needs to allocate (I doubt it), we can't ignore the semantics of `min`/`max` in `extrema`.

N5N3 replied:
Sounds reasonable to me, we can put related change in #41709.

tkf commented on base/reduce.jl:
```diff
@@ -605,67 +608,6 @@ julia> prod(1:5; init = 1.0)
-                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
-    a1 = @inbounds A[first]
-    v1 = mapreduce_first(f, op, a1)
-    v2 = v3 = v4 = v1
-    chunk_len = 256
-    start = first + 1
-    simdstop  = start + chunk_len - 4
-    while simdstop <= last - 3
-        # short circuit in case of NaN or missing
-        (v1 == v1) === true || return v1
-        (v2 == v2) === true || return v2
-        (v3 == v3) === true || return v3
-        (v4 == v4) === true || return v4
-        @inbounds for i in start:4:simdstop
-            v1 = _fast(op, v1, f(A[i+0]))
-            v2 = _fast(op, v2, f(A[i+1]))
-            v3 = _fast(op, v3, f(A[i+2]))
-            v4 = _fast(op, v4, f(A[i+3]))
-        end
-        checkbounds(A, simdstop+3)
-        start += chunk_len
-        simdstop += chunk_len
-    end
-    v = op(op(v1,v2),op(v3,v4))
-    for i in start:last
-        @inbounds ai = A[i]
-        v = op(v, f(ai))
-    end
-
-    # enforce correct order of 0.0 and -0.0
-    # e.g. maximum([0.0, -0.0]) === 0.0
-    # should hold
-    if isbadzero(op, v)
-        for i in first:last
-            x = @inbounds A[i]
-            isgoodzero(op,x) && return x
-        end
-    end
-    return v
-end
```
Why are these methods moved? The commit 54baf1cfac2dd8ed71c7f74affe99dce6d92eaed does not contain an explanation and so it's not clear to me. In general, it'd be nice if you can minimize the diff. Certain code may seem to be out of place but it's much more valuable to keep the Git history understandable. If we really want to change the code organization, I suggest creating an NFC PR that only moves around code without touching the definition.

N5N3 replied:
The new implementation is quite different compared with the current one...
I'll add a NFC commit for easier review.

tkf reviewed (commented):
I left some nitpicky comments for making reviewing easier.

N5N3 committed:
Merge branch 'pr/36265' into PR36265(cleaned)

N5N3 committed:
add test for extrema(!)

N5N3 commented:
Looks like LLVM13 makes Float64 extrema 20% slower on my desktop. (For long inputs)
But `Float32` is much faster:
```julia
julia> a = randn(Float32,4096);
julia> @btime Base.mapreduce_impl(identity,min, $a, 1, 4096)
  493.333 ns (0 allocations: 0 bytes)
-3.9800694f0
julia> @btime Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, 4096)
  533.158 ns (0 allocations: 0 bytes)
(-3.9800694f0, 3.686795f0)
```

N5N3 commented:
After some tuning, LLVM13 is able to optimize `extrema` much better:
```julia
a = randn(T, 4096)
for n = 1:12
    print(1<<n,':')
    t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
    t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
    println("extrema:", round(t1*1e10)/10, "ns  minimum:", round(t2*1e10)/10, "ns")
end
```
`Float64`:
```julia
2:extrema:8.4ns  minimum:7.5ns
4:extrema:11.1ns  minimum:10.0ns
8:extrema:21.1ns  minimum:15.4ns
16:extrema:21.9ns  minimum:14.3ns
32:extrema:28.6ns  minimum:19.3ns
64:extrema:41.6ns  minimum:30.1ns
128:extrema:71.5ns  minimum:47.2ns
256:extrema:130.3ns  minimum:76.2ns
512:extrema:184.7ns  minimum:112.8ns
1024:extrema:294.6ns  minimum:191.2ns
2048:extrema:516.7ns  minimum:352.7ns
4096:extrema:968.4ns  minimum:670.6ns
```
`Float32`:
```julia
2:extrema:7.3ns  minimum:7.3ns
4:extrema:9.5ns  minimum:9.8ns
8:extrema:14.3ns  minimum:15.2ns
16:extrema:16.5ns  minimum:14.1ns
32:extrema:22.6ns  minimum:18.0ns
64:extrema:35.2ns  minimum:28.2ns
128:extrema:51.3ns  minimum:46.5ns
256:extrema:71.4ns  minimum:67.7ns
512:extrema:102.5ns  minimum:95.7ns
1024:extrema:165.7ns  minimum:151.9ns
2048:extrema:292.8ns  minimum:266.7ns
4096:extrema:547.6ns  minimum:495.9ns
```

N5N3 commented:
Update the latest benchmark:

<details><summary>This PR</summary>

```julia
julia> a = randn(Float64,4096);

julia> for n = 1:12
           print(1<<n,':')
           t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
       end
2:extrema:8.4ns  minimum:7.5ns
4:extrema:11.1ns  minimum:10.2ns
8:extrema:21.0ns  minimum:15.7ns
16:extrema:21.2ns  minimum:13.7ns
32:extrema:27.8ns  minimum:18.2ns
64:extrema:40.4ns  minimum:27.8ns
128:extrema:60.0ns  minimum:44.1ns
256:extrema:106.9ns  minimum:70.0ns
512:extrema:154.2ns  minimum:98.9ns
1024:extrema:248.6ns  minimum:157.5ns
2048:extrema:439.9ns  minimum:274.3ns
4096:extrema:820.7ns  minimum:516.7ns

julia> a = randn(Float32,4096);

julia> for n = 1:12
           print(1<<n,':')
           t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
       end
2:extrema:7.5ns  minimum:7.3ns
4:extrema:9.9ns  minimum:10.0ns
8:extrema:14.8ns  minimum:15.8ns
16:extrema:21.0ns  minimum:16.1ns
32:extrema:27.5ns  minimum:21.5ns
64:extrema:40.3ns  minimum:31.6ns
128:extrema:55.5ns  minimum:49.5ns
256:extrema:74.0ns  minimum:73.0ns
512:extrema:100.4ns  minimum:98.2ns
1024:extrema:155.8ns  minimum:148.7ns
2048:extrema:265.6ns  minimum:250.4ns
4096:extrema:485.2ns  minimum:452.0ns
```
</details>


<details><summary>1.7.1</summary>

```julia
julia> a = randn(Float64,4096);

julia> for n = 1:12
           print(1<<n,':')
           #t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           #println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
           println("minimum:", round(1e10t2)/10,"ns")
       end
2:minimum:8.6ns
4:minimum:11.3ns
8:minimum:17.0ns
16:minimum:32.0ns
32:minimum:61.9ns
64:minimum:119.0ns
128:minimum:235.6ns
256:minimum:469.9ns
512:minimum:588.3ns
1024:minimum:837.3ns
2048:minimum:1300.0ns
4096:minimum:2233.3ns

julia> a = randn(Float32,4096);

julia> for n = 1:12
           print(1<<n,':')
           #t1 = @belapsed Base.mapreduce_impl(Base.ExtremaMap(identity),Base._extrema_rf, $a, 1, (1<<$n))
           t2 = @belapsed Base.mapreduce_impl(identity,min, $a, 1, (1<<$n))
           #println("extrema:", round(1e10t1)/10, "ns  minimum:", round(1e10t2)/10,"ns")
           println("minimum:", round(1e10t2)/10,"ns")
       end
2:minimum:9.3ns
4:minimum:12.0ns
8:minimum:17.5ns
16:minimum:31.9ns
32:minimum:61.6ns
64:minimum:132.5ns
128:minimum:400.5ns
256:minimum:918.5ns
512:minimum:991.7ns
1024:minimum:1250.0ns
2048:minimum:1700.0ns
4096:minimum:2555.6ns
```
</details>

tkf commented on base/reduce.jl:
```diff
@@ -785,6 +726,83 @@ Inf
+@inline (f::ExtremaMap)(x) = (y = f.f(x); (y, y))
+
+@inline _extrema_rf((min1, max1), (min2, max2)) = (min(min1, min2), max(max1, max2))
+# optimization for AbstractFloat
+function _extrema_rf(x::NTuple{2,T}, y::NTuple{2,T}) where {T<:AbstractFloat}
+    (x1, x2), (y1, y2) = x, y
+    anynan = isnan(x1)|isnan(y1)
+    z1 = ifelse(anynan, x1-y1, ifelse(signbit(x1-y1), x1, y1))
+    z2 = ifelse(anynan, x1-y1, ifelse(signbit(x2-y2), y2, x2))
+    z1, z2
+end
```
```suggestion
```

IIUC, this is introducing an optimization that is still under discussion in #41709. We can do this after #41709 is merged.

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1264,102 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+function mapreduce_impl(f, op::Union{typeof(max),typeof(min)},
+                        A::AbstractArrayOrBroadcasted, first::Int, last::Int)
+    @inline elf(i::Int) = @inbounds f(A[i])
+    Eltype = _return_type(elf, Tuple{Int})
+    # Limit this optimization to IEEEFloat. (general fallback is 2x faster for Integer)
+    Eltype <: IEEEFloat ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
```
Using the inference API (`_return_type`) like this should probably ensure that the result is bit-wise equal. The "SIMD friendly" implementation below is requiring the communitativity of `min`/`max` but the generic `mapreduce` does not rely on commutativity. This difference can be observed when the user checks the NaN with `===`. If you want to do this, I think using `eltype` seems to be the only reliable choice.

Also, IIUC, this PR's main purpose is to fix the bugs. I personally think separating optimization-only PR is a better approach.

N5N3 replied:
I guess replacing `Eltype <: IEEEFloat` with `isconcretetype(Eltype ) && Eltype <: IEEEFloat` is enough?
It should also solve the following concern.

N5N3 replied:
Seperating seems OK to me, will remove them in this PR.

tkf replied:
No, I don't think it's enough. A subtle change in user code can change the quality of inference of `elf` while preserving the input-output mapping of `elf`. It can happen even if `elf` always returns (say) a `Float64` at run-time.

N5N3 replied:
Do you mean something like `f(x::Float64) = isnan(x) ? 0 : x` while inputs has no NaN?
I guess we can give up optimizion in this case.

N5N3 replied:
I have open #43725 for early review for performance change.
Further discussion could continue there.

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1264,102 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+        while i <= last - N
+            temp = map(elf, i .+ index)
+            _anynan(temp) && return firstnan(temp) # nan check
+            vs = map(_fast(op), vs, temp) # then _fast(op) is safe
```
It seems constructing a tuple like this can hurt performance when we have something like `Eltype === Union{Float32,Float64}`. We either need a manual promotion or restrict `Eltype` to a concrete one.

tkf commented on base/reduce.jl:
```diff
@@ -1246,3 +1264,102 @@ function _simple_count(::typeof(identity), x::Array{Bool}, init::T=0) where {T}
+    Eltype <: IEEEFloat ||
+        return mapreduce_impl(f, op, A, first, last, pairwise_blocksize(f, op))
+    @noinline firstnan(temp::Tuple) = temp[findfirst(isnan, temp)], last
+    function simd_kernal(::Val{N}, ini) where {N}
+        # reused kernal for better vectorizable reduction.
+        # - `N` is the unroll size.
+        # - `ini` is the initial value of reduction
+        # It returns the reduced result (`v`), and the last processed index (`i`).
+        isnan(ini) && return ini, last
+        vs = ntuple(Returns(ini), Val(N))
+        index = ntuple(identity, Val(N))
+        local i = first
+        while i <= last - N
+            temp = map(elf, i .+ index)
+            _anynan(temp) && return firstnan(temp) # nan check
+            vs = map(_fast(op), vs, temp) # then _fast(op) is safe
+            i += N
+        end
+        _half_reduce(_fast(op), vs), i
+    end
```
I know that the pre-PR version also had the manual unrolling, but if we are doing this I wonder if it's better to just ask LLVM...

@vchuravy Which one do you hate less, manual unrolling vs `Expr(:loopinfo, ...)`? Or is there a better way?

N5N3 replied:
It's should be easy to use `Expr(:loopinfo, ...)` to realize `vs = map(_fast(op), vs, temp)`, 
But I guess thats not true for `_anynan(temp) && return firstnan(temp)`?

tkf commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -1914,13 +1916,16 @@ function Base._mapreduce(f, op, ::Base.IndexCartesian, A::AbstractSparseMatrixCS
+    nzeros == 0 ? op(oftype(v0, 0), v0) : op(f(zero(T))*nzeros, v0)
 _mapreducezeros(f, ::typeof(*), ::Type{T}, nzeros::Integer, v0) where {T} =
     nzeros == 0 ? v0 : f(zero(T))^nzeros * v0
+_mapreducezeros(f, op::Union{typeof(min),typeof(max),typeof(Base._extrema_rf)},
+                                ::Type{T}, nzeros::Integer, v0) where {T} =
+    nzeros == 0 ? v0 : op(v0, f(zero(T)))
```
This is great!

N5N3 committed:
Remove `_extrema_rf` optimization for `AbstractFloat` (follow advise)

Add TODO

apply suggestions

Update NEWS.md

Co-Authored-By: Takafumi Arakaki <takafumi.a@gmail.com>
Co-Authored-By: Takafumi Arakaki <29282+tkf@users.noreply.github.com>

N5N3 commented:
Performance related changes have been removed. Will open a seperate PR, once this is merged.

N5N3 committed:
remove short circuit for correctness

This optimization gives wrong result, if `NaN` and `missing` both exist.

add missing test

tkf commented on base/reducedim.jl:
```diff
@@ -77,21 +77,21 @@ end
 reducedim_initarray(A::AbstractArrayOrBroadcasted, region, init, ::Type{R}) where {R} = fill!(similar(A,R,reduced_indices(A,region)), init)
 reducedim_initarray(A::AbstractArrayOrBroadcasted, region, init::T) where {T} = reducedim_initarray(A, region, init, T)
-
+reducedim_initarray(A::AbstractArrayOrBroadcasted, region, ::UndefInitializer, ::Type{R}) where {R} = similar(A,R,reduced_indices(A,region))
```
Use `_InitialValue` instead of `UndefInitializer`. It is perfectly OK for a user to implement a monoid involving `UndefInitializer` and passing `init = undef` because `undef` is a public API.

tkf commented on base/reducedim.jl:
```diff
@@ -125,46 +125,38 @@ function _reducedim_init(f, op, fv, fop, A, region)
-            # v0 may have changed type.
-            Tr = v0 isa T ? T : typeof(v0)
+    # calculate the output type
+    T = promote_typejoin_union(_return_type(f, Tuple{eltype(A)}))
```
Please don't use `_return_type` in a way that the user can observe the precision of the inference.

N5N3 replied:
Ummm, but we need it to know the `eltype` of output array.
Is there a better way to do this?

N5N3 replied:
In fact the following code cause crash on 1.7:
```julia
julia> a = collect(1.0:4);

julia> f(x) = x < 3 ? missing : x;

julia> minimum(f, a; dims = 1)
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives
<unknown>:0: error: starting new .cfi frame before finishing the previous one
<unknown>:0: error: this directive must appear between .cfi_startproc and .cfi_endproc directives

Please submit a bug report with steps to reproduce this fault, and any error messages that follow (in their entirety). Thanks.
Exception: EXCEPTION_ACCESS_VIOLATION at 0x0 -- unknown function (ip: 0000000000000000)
in expression starting at none:0
unknown function (ip: 0000000000000000)
warning: DWARF compile unit extends beyond its bounds cu 0x00000000 at 0x0000000b

error: No unit DIE
unknown function (ip: 000000004eaf2fb7)
#current_exceptions#67 at .\error.jl:155
current_exceptions at .\error.jl:149 [inlined]
current_exceptions at .\error.jl:149 [inlined]
eval_user_input at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:162
repl_backend_loop at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:244
start_repl_backend at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:229
#run_repl#47 at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:362
run_repl at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.7\REPL\src\REPL.jl:349
#930 at .\client.jl:394
jfptr_YY.930_35191.clone_1 at C:\Users\MYJ\AppData\Local\Programs\Julia-1.7.1\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1788 [inlined]
jl_f__call_latest at /cygdrive/c/buildbot/worker/package_win64/build/src\builtins.c:757
#invokelatest#2 at .\essentials.jl:716 [inlined]
invokelatest at .\essentials.jl:714 [inlined]
run_main_repl at .\client.jl:379
exec_options at .\client.jl:309
_start at .\client.jl:495
jfptr__start_43221.clone_1 at C:\Users\MYJ\AppData\Local\Programs\Julia-1.7.1\lib\julia\sys.dll (unknown line)
jl_apply at /cygdrive/c/buildbot/worker/package_win64/build/src\julia.h:1788 [inlined]
true_main at /cygdrive/c/buildbot/worker/package_win64/build/src\jlapi.c:559
jl_repl_entrypoint at /cygdrive/c/buildbot/worker/package_win64/build/src\jlapi.c:701
mainCRTStartup at /cygdrive/c/buildbot/worker/package_win64/build/cli\loader_exe.c:42
BaseThreadInitThunk at C:\WINDOWS\System32\KERNEL32.DLL (unknown line)
RtlUserThreadStart at C:\WINDOWS\SYSTEM32\ntdll.dll (unknown line)
Allocations: 2401717 (Pool: 2400926; Big: 791); GC: 3
```

N5N3 replied:
I guess #35323 is able to fix the crash.
But it still throw error unneeded for the following
```julia
julia> a = range(.0,1,4);

julia> f(x) = x < 0.5 ? 1 : x;

julia> minimum(f, a; dims = 1)
ERROR: InexactError: Int64(0.6666666666666666)
```
If `f` is unstable, I guess `Array{Real}` or `Array{Any}` is better than a `ERROR`?

N5N3 replied:
I try to look into the histroy of preallocation in `Base`.
Looks like inference based initialization is prefered only for empty cases.
But for `mapreduce`, if we alow user to change the elements type of inputs (`sum(Float64, [1,2,3]; dims = 1)`).
I guess we'd better not only use "the first slice" to determine the output eltype, as any unstable `f` might cause problem.
For example:
```julia
julia> sum(x -> x < 0.5 ? 0 : x, range(0.0,1,4), dims = 1)
ERROR: InexactError: Int64(1.6666666666666665)
```
User (like me) might argue that `Vector{Real}` or `Vector{Float64}` is a better choise for preallocation.
I saw #29739 removes
 > !!! warning
    In pathological cases, the type returned by `promote_op(f, argtypes...)` may not even
    be a supertype of the return value of `f(::argtypes...)`. Therefore, `promote_op`
    should _not_ be used e.g. in the preallocation of an output array.

@vtjnash is our inference system more reliable today?

N5N3 replied:
@tkf. Inference based preallocation is limited to concrete cases. Otherwise use the current logic to derive eltype.
I'm not sure if this is better.

tkf replied:
If you check if `_return_type` returns a concrete type, you are leaking the implementation detail of the compiler that the given function can be inferred to return a concrete type. It is true that Base and stdlib have a lot of usages of the compiler API that leak what the compiler does. It may be even possible to argue that this is a necessary evil and removing such usages would have delayed Julia 1.0 release by years. But I just don't want to contribute to increasing invalid usages myself (at least until all other options are explored).

> @vtjnash is our inference system more reliable today?

This question is irrelevant in this context. Correct Julia code does not rely on what the compiler does.

N5N3 replied:
Well then I'll use the current code for initialization, and mark some test as broken.

N5N3 commented on base/reduce.jl:
```diff
@@ -634,11 +637,6 @@ function mapreduce_impl(f, op::Union{typeof(max), typeof(min)},
     start = first + 1
     simdstop  = start + chunk_len - 4
     while simdstop <= last - 3
-        # short circuit in case of NaN or missing
-        (v1 == v1) === true || return v1
-        (v2 == v2) === true || return v2
-        (v3 == v3) === true || return v3
-        (v4 == v4) === true || return v4
```
This is definitely incorrect if inputs contains both `missing` and `NaN`. (It might return a `NaN`, but we should always return `missing`). Related test is added in the same commit.

N5N3 committed:
API clean

and export `extrema!`

fix doc

Update collections.md

drop sparsearray change

Co-Authored-By: Takafumi Arakaki <29282+tkf@users.noreply.github.com>

N5N3 commented:
`SparseArrays` related change has been moved to solve confict. But I guess CI will fail for sparsearray tests as some errors' type is changed.
@DilumAluthge Is there a better way to solve the problem?

tkf commented on base/reduce.jl:
```diff
@@ -439,6 +439,9 @@ mapreduce(f, op, a::Number) = mapreduce_first(f, op, a)
 
 _mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f, op, A)
 
+@inline _mapreduce(f, op, ::IndexCartesian, A::AbstractVector) =
+    _mapreduce(f, op, IndexLinear(), A)
```
We can handle 1D Broadcasted too

```suggestion
function _mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted)
    if IteratorSize(A) isa HasShape
        if ndims(A) == 1
            return _mapreduce(f, op, IndexLinear(), A)
        end
    end
    return mapfoldl(f, op, A)
end
```

N5N3 replied:
I guess we don't need this. As all one-dimensional `Broadcasted` is linear-indexed if their `ndims` has been defined.
Only `AbstractVector` might been cartesian-indexed.
https://github.com/JuliaLang/julia/blob/a32a0660dfe28442b3b3e823021462ebbeb90729/base/broadcast.jl#L242-L248

tkf replied:
I think it'd be more productive to mark the discussion resolved once we arrive at a conclusion.

Notice that your reasoning requires that `mapfoldl(f, op, A::Broadcasted)` dispatches to a `getindex`-based implementation. I just checked that this is not the case with

```julia
julia> bc = Broadcast.instantiate(Broadcast.broadcasted(+, [1,2,3], 4));

julia> @descend mapfoldl(identity, +, bc)
```

N5N3 replied:
Sorry for my recklessness. But I'm affraid I didn't catch you.
For your example `IndexStyle(bc) === IndexLinear()`, thus it won't hit `_mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted)`. Am I right?

tkf replied:
Thanks for the clarification. So you were saying that it's the caller who specifies the correct index style for `bc` with `ndim(bc) == 1`. Yes, indeed that seems to be the case.

But then this raises another question: why do we need to enforce the index style of `_mapreduce(f, op, ::IndexCartesian, A::AbstractVector)`? Arguably, it is possible that `typeof(A)` provides an efficient `iterate` than `getindex` which is why it's not declared to be `IndexLinear` in the first place. I'm not sure if SparseArrays.jl does this but I'd guess you can build a state machine for `iterate` on a sparse vector that is faster than `getindex`. It applies to other non-trivial vector types too. If we call `IndexLinear` here, we are disabling the optimization done by the type implementer.

So, my suggestion now is to simply remove `_mapreduce(f, op, ::IndexCartesian, A::AbstractVector)`

```suggestion
_mapreduce(f, op, ::IndexCartesian, A::AbstractArrayOrBroadcasted) = mapfoldl(f, op, A)
```

N5N3 replied:
I do the above change mainly because our `fold` messes up at speed some time.
```julia
julia> a = randn(4096);
julia> @btime foldl(Base.add_sum,$a);
  3.688 μs (0 allocations: 0 bytes)
julia> @btime reduce(Base.add_sum,$a);
  296.935 ns (0 allocations: 0 bytes)
```
> Arguably, it is possible that typeof(A) provides an efficient iterate than getindex which is why it's not declared to be IndexLinear in the first place.

Our user defined vectors are default to be `IndexCartesian`.
And it's also a little confused to me that `IndexCartesian` means `iterate` is more efficient than `getindex`.
Anyway, I'll remove it.

N5N3 committed:
Re-implement `reducedim_init` for extrema

N5N3 committed:
Test fix

Mark `BigInt` as broken

N5N3 committed:
Merge branch 'master' into extrama

N5N3 committed:
Inference fix

Update reducedim.jl

tkf commented on base/reducedim.jl:
```diff
@@ -165,6 +164,41 @@ for (f1, f2, initval, typeextreme) in ((:min, :max, :Inf, :typemax), (:max, :min
+        v0 = oftype(v0[1], Inf), oftype(v0[2], -Inf)
+    elseif isunordered(v0[1])
+        # v0 is missing or a third-party unordered value
+        T1, T2 = Tr.parameters[1], Tr.parameters[2]
```
It's better to avoid touching the internal presentation of `DataType`. Why not

```suggestion
        T1 = typeof(Tr[1])
        T1 = typeof(Tr[2])
```

?

N5N3 replied:
I guess `getindex` is undefined for `Tuple`?
```julia
Tuple{Int,Int}[1]
ERROR: MethodError: Cannot `convert` an object of type Int64 to an object of type Tuple{Int64, Int64}
```
I'll split them at the beginning to avoid accessing `parameters`

tkf replied:
Right, I meant something like

```suggestion
        T1 = typeof(v0[1])
        T1 = typeof(v0[2])
```

?

N5N3 replied:
For this case, both `T1` and `T2` are `Missing` . I guess the best choice is the following:
```julia
    Tmin = v0[1] isa T ? T : typeof(v0[1])
    Tmax = v0[2] isa T ? T : typeof(v0[2])
    if v0[1] isa Number && isnan(v0[1])
        v0 = oftype(v0[1], Inf), oftype(v0[2], -Inf)
    elseif isunordered(v0[1])
        v0 = typemax(nonmissingtype(Tmin)), typemin(nonmissingtype(Tmax))
    end
```

tkf replied:
Yes, that looks like the closest to how `minimum`/`maximum` handles this

tkf reviewed (commented):
LGTM! (other than one nitpick and that we need to update SparseArrays)

tkf commented:
I followed https://julialang.github.io/BumpStdlibs.jl/dev/usage/ and triggered the bump https://github.com/JuliaLang/BumpStdlibs.jl/runs/4835402987?check_suite_focus=true

N5N3 commented:
Local test shows no problem
```julia
$ make -C test SparseArrays
make: Entering directory '/cygdrive/c/Users/MYJ/Documents/GitHub/julia/test'
    JULIA test/SparseArrays
Test                    (Worker) | Time (s) | GC (s) | GC % | Alloc (MB) | RSS (MB)
SparseArrays/sparsevector    (4) |        started at 2022-01-17T01:38:57.142
SparseArrays/sparse          (3) |        started at 2022-01-17T01:38:57.262
SparseArrays/higherorderfns  (2) |        started at 2022-01-17T01:38:57.358
SparseArrays/higherorderfns  (2) |    94.64 |   2.66 |  2.8 |   11985.15 |   562.00
SparseArrays/sparsevector    (4) |   164.82 |   4.87 |  3.0 |   24421.82 |   876.65
SparseArrays/sparse          (3) |   263.28 |  23.08 |  8.8 |   27776.47 |   953.35

Test Summary: |  Pass  Broken  Total     Time
  Overall     | 21792       5  21797  4m26.2s
    SUCCESS
make: Leaving directory '/cygdrive/c/Users/MYJ/Documents/GitHub/julia/test'
```
The above suggestion will be adopted together with the actual update of `SparseArrays`.

N5N3 committed:
Merge branch 'master' into extrama

tkf reviewed (approved):
LGTM! Thanks for addressing a lot of fix requests!

greimel commented:
Sorry if it's not appropriate to comment on this long-merged PR.

In the docstring it says

> The value returned for empty `itr` can be specified by `init`. It must be a 2-tuple whose
 first and second elements are neutral elements for `min` and `max` respectively
 (i.e. which are greater/less than or equal to any other element).

However, one very important use-case of this PR could be to compute extrema iteratively (e.g. update bounds as new data come in).

```julia
julia> x = 0:10
0:10

julia> y = -1:5
-1:5

julia> extrema(x)
(0, 10)

julia> extrema(y)
(-1, 5)

julia> extrema(y, init=extrema(x)) == extrema(x ∪ y) == (-1, 10)
true
```

The docstring makes me wonder if this is a valid use of the function, because it says `(i.e. which are greater/less than or equal to any other element)`. That is, can I be sure that this condition will be checked in the future and give an error if it's not satisfied?

Would you be open to a PR that adds my example to the docstring and the tests?

N5N3 commented:
I believe @tkf followed minimum/maximum's docstring, as for `maximum`
```
  The value returned for empty itr can be specified by init. It must be a neutral element for max (i.e. which is less than or equal to any other element) as it is       
  unspecified whether init is used for non-empty collections.
```
And we have the following in `sum`'s docstring
```
  The value returned for empty itr can be specified by init. It must be the additive identity (i.e. zero) as it is unspecified whether init is used for non-empty        
  collections.
```

However, your use case is valid (and intuitive) under current implemenation.
If we want to clarify it, then we'd better also modify other reductions' docstring to keep consistancy.
Anyway, PR (with goodwill) is always welcome.

greimel commented:
Great! I've prepared #44819 with updates for `minimum`, `maximum` and `extrema`.
